### ISSUE TITLE
no_std, no alloc serialization

### ISSUE THREAD

**Issue by kolegs** [capnproto/capnproto-rust#449]
Hi,

I am trying to serialize the data in a highly restricted environment. It is the first time I am using Cap'N proto and I got some troubles. I would like to use `serialize_packed` but I see that `write_message` is marked as only available if `alloc` is configured.

Any hints how to do it? What I would like to achieve is to somehow get access to raw bytes.

I have defined a small struct as a test:
```
struct TestData {
    foo @0 :UInt8;
    bar @1 :UInt16;
}
```
And I tried to use it this way:
```
#[cfg(test)]
mod tests {
    use crate::frame_capnp;
    use capnp::message::{Builder, ReaderSegments, SingleSegmentAllocator};
    use capnp::serialize_packed;

    #[test]
    fn test() {
        let mut buffer = [capnp::word(0, 0, 0, 0, 0, 0, 0, 0); 256];
        let allocator =
            SingleSegmentAllocator::new(capnp::Word::words_to_bytes_mut(&mut buffer[..]));

        let mut message = Builder::new(allocator);

        let mut td = message.init_root::<frame_capnp::test_data::Builder>();
        td.set_foo(1);
        td.set_bar(2);

        let size = td.total_size().unwrap();
        assert_eq!(1, size.word_count);

        let s = td.into_reader();
        assert_eq!(1, s.get_foo());
        assert_eq!(2, s.get_bar());

        // How to get raw bytes here? and packed if possible
    }
}

```

**Comment by dwrensha** [comment]
You can get the bytes by calling `message.get_segments_for_output()`, which will give you a single-segment `OutputSegments`.

Currently there is no way to use the packed encoding in no-alloc mode. I think we could add support for that and for plain `serialize_packed::write_message()` by putting the [multi-segment support](https://github.com/capnproto/capnproto-rust/blob/master/capnp/src/serialize.rs#L508-L514) behind a `#[cfg]` attribute.

**Comment by kolegs** [comment]
Thanks for help.
Is adding support for packed encoding in no-alloc mode something I could help with?

**Comment by dwrensha** [comment]
The big obstacle I see is on the reading side of things. Currently `serialize_packed::read_message()` allocates a vector on the heap: https://github.com/capnproto/capnproto-rust/blob/18d362fadbd25413759c04656767da22b49bf4e0/capnp/src/serialize.rs#L239

It's unclear to me how we should make that work in the no-alloc case. I suppose we need to add a way to make that allocation configurable, but I don't see an obvious nice way to do it.

**Comment by kolegs** [comment]
Usually in the situations like that the user should decide of the size of the buffer used to serialize/deserialize data. 
I used in the past this library for the c++/protobuf: https://embeddedproto.com/ and this is written in an elegant way. But using the same approach would require a lot of work here. And of course this library is only targeting embedded development.

**Comment by dwrensha** [comment]
We could add a `serialize::read_message_into_buffer_segments()` function which is like `serialize::read_message()`, but additionally takes a `&[u8]` to use as the backing storage, and returns a `message::Reader<BufferSegments<&[u8]>>` (instead of a `message::Reader<OwnedSegments>`).

**Comment by kolegs** [comment]
I can try to play a little bit with it in the upcoming days if you want, but be aware I am not very familiar with the code and CapN proto library.

Off topic question:
I'm testing the library a little bit and I added `AnyPointer` and `List(Uint16)` to one struct. But when I call `init_` method to initialize the field the struct itself is borrowed so I cannot make another call to initialize another field. Is this a limitation? Of is there another way?

**Comment by dwrensha** [comment]
> But when I call init_ method to initialize the field the struct itself is borrowed so I cannot make another call to initialize another field. 

You cannot have a mutable reference to both child fields at the same time, but you can initialize them sequentially using the `reborrow()` method: https://dwrensha.github.io/capnproto-rust/2014/12/27/custom-mutable-references.html

**Comment by dwrensha** [comment]
#451 adds no-alloc `serialize_packed::write_message()` and `serialize_packed::read_message_no_alloc()`.

**Comment by dwrensha** [comment]
Fixed in #451, and released in [capnp-v0.18.5](https://crates.io/crates/capnp/0.18.5).
